devenv/lib/archive.py

```diff
@@ -21,23 +22,24 @@ def download(url: str, sha256: str, dest: str = "") -> str:
             resp = urllib.request.urlopen(url)
         except HTTPError as e:
             raise RuntimeError(f"Error getting {url}: {e}")
-        with open(dest, "wb") as f:
-            shutil.copyfileobj(resp, f)
-
-    checksum = hashlib.sha256()
-    with open(dest, mode="rb") as f:
-        bts = f.read(4096)
-        while bts:
-            checksum.update(bts)
-            bts = f.read(4096)
-
-    if not secrets.compare_digest(checksum.hexdigest(), sha256):
-        os.remove(dest)
-        raise RuntimeError(
-            f"checksum mismatch for {url}:\n"
-            f"- got: {checksum.hexdigest()}\n"
-            f"- expected: {sha256}\n"
-        )
+
+        with tempfile.NamedTemporaryFile(delete=False) as tmpf:
```

Comment:
specifically this tempfile has to be adjacent to where it will end up -- otherwise it could attempt to rename across partitions -- you can use `dir=...` to control that

Reply:
oh, i suppose rename atomicity is only guaranteed within the same fs, that makes sense

Reply:
yeah -- it's even an error when attempting a rename across filesystems